src/lib/logical_query_plan/join_node.cpp

```diff
@@ -92,35 +101,89 @@ std::shared_ptr<LQPUniqueConstraints> JoinNode::unique_constraints() const {
+  bool left_operand_is_unique =
+      !left_unique_constraints->empty() &&
+      contains_matching_unique_constraint(left_unique_constraints, {join_predicate->left_operand()});
+  bool right_operand_is_unique =
+      !right_unique_constraints->empty() &&
+      contains_matching_unique_constraint(right_unique_constraints, {join_predicate->right_operand()});
 
   if (left_operand_is_unique && right_operand_is_unique) {
     // Due to the one-to-one relationship, the constraints of both sides remain valid.
-    for (const auto& unique_constraint : *left_unique_constraints) {
-      unique_constraints->emplace_back(unique_constraint);
-    }
-    for (const auto& unique_constraint : *right_unique_constraints) {
-      unique_constraints->emplace_back(unique_constraint);
-    }
+    auto unique_constraints =
+        std::make_shared<LQPUniqueConstraints>(left_unique_constraints->begin(), left_unique_constraints->end());
+    std::copy(right_unique_constraints->begin(), right_unique_constraints->end(),
+              std::back_inserter(*unique_constraints));
+    return unique_constraints;
 
   } else if (left_operand_is_unique) {
     // Uniqueness on the left prevents duplication of records on the right
     return right_unique_constraints;
-
   } else if (right_operand_is_unique) {
     // Uniqueness on the right prevents duplication of records on the left
     return left_unique_constraints;
   }
 
-  return unique_constraints;
+  return std::make_shared<LQPUniqueConstraints>();
+}
+
+std::vector<FunctionalDependency> JoinNode::non_trivial_functional_dependencies() const {
+  /**
+   * In the case of Semi- & Anti-Joins, this node acts as a filter for the left input node. The number of output
+   * expressions does not change and therefore we should forward non-trivial FDs as follows:
+   */
+  if (join_mode == JoinMode::Semi || join_mode == JoinMode::AntiNullAsTrue || join_mode == JoinMode::AntiNullAsFalse) {
+    return left_input()->non_trivial_functional_dependencies();
+  }
+
+  auto fds_left = left_input()->non_trivial_functional_dependencies();
+  auto fds_right = right_input()->non_trivial_functional_dependencies();
+
+  /**
+   * When joining tables, we usually lose some or even all unique constraints from both input tables. This leads to
+   * less trivial FDs that we can generate from unique constraints in upper nodes. To preserve all FDs possible, we
+   * generate and return all FDs from both input nodes.
+   * In the following, we determine the unique constraints that become discarded and built FDs from them.
+   */
+  const auto& left_unique_constraints = left_input()->unique_constraints();
+  const auto& right_unique_constraints = right_input()->unique_constraints();
+  const auto& valid_unique_constraints = _valid_unique_constraints(left_unique_constraints, right_unique_constraints);
+  if (valid_unique_constraints->empty()) {
+    // All unique constraints become discarded
+    if (!left_unique_constraints->empty()) {
```

Comment:
Do we need the `x->empty()` checks? `merge_fds` has a shortcut and I don't think they add anything to the clarity here.

Reply:
Yes, it is necessary to fulfill an Assert in `fds_from_unique_constraints()`.

But, I found a better solution:

```cpp
if (valid_unique_constraints->empty()) {
    // All unique constraints become discarded, so we have to manually forward all FDs from both input nodes.
    fds_left = left_input()->functional_dependencies();
    fds_right = right_input()->functional_dependencies();
  } else if (valid_unique_constraints == right_unique_constraints) {
    // Left input node's unique constraints become discarded
    fds_left = left_input()->functional_dependencies();
    fds_right = right_input()->non_trivial_functional_dependencies();
  } else if (valid_unique_constraints == left_unique_constraints) {
    // Right input node's unique constraints become discarded
    fds_left = left_input()->non_trivial_functional_dependencies();
    fds_right = right_input()->functional_dependencies();
  } else {
    // All unique constraints from input nodes remain valid. Therefore, we only have to forward the non-trivial FDs.
    DebugAssert(
        valid_unique_constraints->size() == (left_unique_constraints->size() + right_unique_constraints->size()),
        "Unexpected number of unique constraints.");
    fds_left = left_input()->non_trivial_functional_dependencies();
    fds_right = right_input()->non_trivial_functional_dependencies();
  }
```

Looks more intuitive to me.